File: leetcode/FlattenBinaryTree toLinkedList.cpp
/*
time   :  2015-5-16
author :  ht
comment:  将一棵二叉树变为只有右子树的树，类似一条链表
//题目说了，需要原址重排，不能改变root的地址，改了也没用
*/
#include<iostream>
#include<cmath>
#include<cstring>

using namespace std;


//Definition for a binary tree node.
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
};

TreeNode* Node(TreeNode* current, int *index, int*value, int nums)
{
	//这个函数用来创建树 创建方式是先序遍历
	if (nums <= 0)
		return NULL;
	current = (TreeNode*)malloc(sizeof(TreeNode));
	if (value[index[0]] != -1)
	{
		current->val = value[index[0]];
		current->left = NULL;
		index[0]++;
		if (index[0] < nums)
		{
			current->left = Node(current->left, index, value, nums);
		}
		current->right = NULL;
		if (index[0] < nums)
		{
			current->right = Node(current->right, index, value, nums);
		}
	}
	else
	{
		current = NULL;
		index[0]++;
	}
	return current;
}

int traverse(struct TreeNode* current, int *index, int* result, int depth)
{
	if ((current->left == NULL) && (current->right == NULL))
	{

		return 0;
	}
	else
	{
		struct TreeNode* left = NULL, *right = NULL,*p=NULL;
		if (current->left != NULL)
		{
			left = current->left;
			traverse(current->left, index, result, depth + 1);
			if (current->right != NULL)
			{
				current->left = NULL;
				
				right = current->right;
				
				current->right = left;
				
				p = left;
				//关键点在这个循环上，因为当左子树添加到右子树上的时候
				//必须找到左子树的最右节点，然后最右节点链接到以前的右子树上
				while (p->right != NULL)
					p = p->right;


				p->right = right;
				traverse(right, index, result, depth + 1);
				return 1;
			}
			//右边是空的直接把左边的接上去就行了
			current->left = NULL;
			current->right = left;
			return 2;
			
		}
		else
		{
			//左子树是空，什么也不用做，直接遍历右子树即可
			traverse(current->right, index, result, depth + 1);
			return 3;  //其实返回值没有多大用
		}
		
	}
}
void flatten(struct TreeNode* root) {
	if (root == NULL)
		return;
	int index = 0;
	int size = 10000;
	int *result = (int*)malloc(size*sizeof(int));
	int depth = traverse(root, &index, result, 0);
}

int main()
{
	//int value[6] = { 1,-1,2,3,-1,2};
	//int value[13] = { 1, 2, -1, -1, 4, 5, 3, -1, -1, 8, -1, -1, -1 };
	//int value[15] = { 1, 2, 4, -1, -1, 5, -1, -1, 3, 6, -1, -1, 7, -1, -1 };
	//这是一颗满树
	int value[7] = { 1,2,4,-1,-1,-1,3};
	//int value[6] = { 1, 2, -1, -1, -1 };
	TreeNode* root = NULL, *p = NULL;
	int index = 0;
	int nums = 7;
	root = Node(root, &index, value, nums);

	int returnSize = 0;
	flatten(root);
	return 0;
}

File: leetcode/src/FlattenBinaryTreetoLinkedList.java
/**
 * Created by hduser on 15-12-4.
 * Given a binary tree, flatten it to a linked list in-place.

 For example,
 Given

     1
    / \
   2   5
  / \   \
 3   4   6
 The flattened tree should look like:
 1
 \
 2
 \
 3
 \
 4
 \
 5
 \
 6
 If you notice carefully in the flattened tree, each node's right child points to the next node of a pre-order traversal.
 */
import ht.TreeNode;
public class FlattenBinaryTreetoLinkedList {
    public void flatten(TreeNode root) {
        helper(root);
    }
    private void helper(TreeNode root) {
        //TreeNode head = new TreeNode(Integer.MAX_VALUE);
        //head.left = root;
        //TreeNode tail = new TreeNode(Integer.MIN_VALUE);
        //head = dfs(head, tail);
        //root = head.right;
        //TreeNode[] label = dfs2(head);
        //root = head.right;
        TreeNode[] tail = new TreeNode[1];
        tail[0] = null;
        dfs3(root, tail);
        return;
    }
    private TreeNode dfs(TreeNode root, TreeNode tail) {
        if (root.right == null && root.left == null) {
            tail.right = root;
            return root;
        } else {
            if (root.left == null || root.right == null) {
                root.left = root.right == null ? root.left : root.right;
                TreeNode head = dfs(root.left, tail);
                root.right = head;
                root.left = null;
                return root;
            } else {
                TreeNode current = root.right;
                TreeNode head = dfs(root.left, tail);
                root.right = head;
                root.left = null;

                TreeNode tmpTail = new TreeNode(Integer.MIN_VALUE);
                head = dfs(current, tmpTail);
                tail.right.right = head;
                tail.right = tmpTail.right;
                return root;
            }
        }
    }
    private TreeNode[] dfs2(TreeNode root) {
        // 每次递归返回首尾节点
        if (root.right == null && root.left == null) {
            TreeNode[] label = {root, root};
            return label;
        } else {
            if (root.left == null || root.right == null) {
                root.left = root.right == null ? root.left : root.right;
                TreeNode[] leftLabel = dfs2(root.left);
                root.right = leftLabel[0];
                root.left = null;
                leftLabel[0] = root;
                return leftLabel;
            } else {
                TreeNode current = root.right;
                TreeNode[] leftLabel = dfs2(root.left);
                root.right = leftLabel[0];
                root.left = null;

                TreeNode[] rightLabel = dfs2(current);
                leftLabel[1].right = rightLabel[0];
                rightLabel[0] = root;
                return rightLabel;
            }
        }
    }
    private TreeNode dfs3(TreeNode root, TreeNode[] tail) {
        // 这个递归是最简洁的
        if (root == null) {
            return tail[0];
        }
        TreeNode[] head = new TreeNode[1];
        head[0] = dfs3(root.right, tail);
        root.right = dfs3(root.left, head);
        root.left = null;
        return root;
    }
}
